### ISSUE TITLE
Temporary filename collision

### ISSUE THREAD

**Issue by justinormont** [dotnet/machinelearning#568]
### System information

- **OS version/distro**: Windows Server 2016

### Issue

When running simultaneous sweeps (multiple processes), one process tried to open a new temp file which already exists. 

`Unexpected exception: The file 'C:\Users\jormont\AppData\Local\Temp\2\TLC_25553D77\0' already exists`

- **What did you do?**
Running an AutoML macro sweep w/ four processes on the same machine. 

Perhaps our temporary path names aren't unique enough? 

https://github.com/dotnet/machinelearning/blob/211c0439aa7bec33747536a273f90b1718ead24a/src/Microsoft.ML.Data/Model/Repository.cs#L507

Unsure which call is used to create the `TLC_...` folder, but could be this one:
https://github.com/dotnet/machinelearning/blob/c023727b76970ab913ec1ce38276508835c17bcf/src/Microsoft.ML.Core/Environment/TlcEnvironment.cs#L398

### Source code / logs

Error:
```
Unexpected exception: The file 'C:\Users\jormont\AppData\Local\Temp\2\TLC_25553D77\0' already exists., 'System.IO.IOException'
   at System.IO.__Error.WinIOError(Int32 errorCode, String maybeFullPath)
   at System.IO.FileStream.Init(String path, FileMode mode, FileAccess access, Int32 rights, Boolean useRights, FileShare share, Int32 bufferSize, FileOptions options, SECURITY_ATTRIBUTES secAttrs, String msgPath, Boolean bFromProxy, Boolean useLongPath, Boolean checkHost)
   at System.IO.FileStream..ctor(String path, FileMode mode, FileAccess access, FileShare share)
   at System.IO.Compression.ZipFileExtensions.ExtractToFile(ZipArchiveEntry source, String destinationFileName, Boolean overwrite)
   at Microsoft.ML.Runtime.Model.RepositoryReader.OpenEntryOrNull(String dir, String name) in D:\TLC\OpenSource\src\Microsoft.ML.Data\Model\Repository.cs:line 507
   at Microsoft.ML.Runtime.Model.ModelLoadContext.LoadModelOrNull[TRes,TSig](IHostEnvironment env, TRes& result, RepositoryReader rep, String dir, Object[] extra) in D:\TLC\OpenSource\src\Microsoft.ML.Data\Model\ModelLoading.cs:line 55
   at Microsoft.ML.Runtime.Model.ModelLoadContext.LoadModel[TRes,TSig](IHostEnvironment env, TRes& result, RepositoryReader rep, String dir, Object[] extra) in D:\TLC\OpenSource\src\Microsoft.ML.Data\Model\ModelLoading.cs:line 89
   at Microsoft.ML.Runtime.Data.ApplyTransformUtils.ApplyTransformToData(IHostEnvironment env, IDataTransform transform, IDataView newSource) in D:\TLC\OpenSource\src\Microsoft.ML.Data\Utilities\ApplyTransformUtils.cs:line 47
   at Microsoft.ML.Runtime.Data.ApplyTransformUtils.ApplyAllTransformsToData(IHostEnvironment env, IDataView chain, IDataView newSource, IDataView oldSource) in D:\TLC\OpenSource\src\Microsoft.ML.Data\Utilities\ApplyTransformUtils.cs:line 101
   at Microsoft.ML.Runtime.EntryPoints.TransformModel..ctor(IHostEnvironment env, IDataView result, IDataView input) in D:\TLC\OpenSource\src\Microsoft.ML.Data\EntryPoints\TransformModel.cs:line 62
   at Microsoft.ML.Runtime.Data.TextAnalytics.TextTransform(IHostEnvironment env, Arguments input) in D:\TLC\source\TextAnalytics\TextTransform.cs:line 549
```

**Comment by jwood803** [comment]
@justinormont @TomFinley Tried to take a look at this, but I'm wondering if the referenced code isn't there anymore.

Did a quick search and found this:
https://github.com/dotnet/machinelearning/blob/e0c4caae69735b9ee2df3ede44cd7a5ef29ff681/src/Microsoft.ML.Core/ComponentModel/AssemblyLoadingUtils.cs#L106

It looks like it's using a GUID. If this is the correct code, would this be considered fixed already?

**Comment by justinormont** [comment]
@jwood803, thanks for taking a look.

The referenced code moved named spaces to here:
https://github.com/dotnet/machinelearning/blob/37ed3361f330ed439c9bd477abcfe5a27e06c0a1/src/Microsoft.ML.Core/Data/Repository.cs#L527

The `pathTemp` is created in `GetPath()`:
https://github.com/dotnet/machinelearning/blob/37ed3361f330ed439c9bd477abcfe5a27e06c0a1/src/Microsoft.ML.Core/Data/Repository.cs#L242-L282

**Comment by jwood803** [comment]
@justinormont Apologies, I totally forgot that I had started looking into this.

Looking again and, from what I can tell, it just needs to check if the temp path already exists and, if it does, make sure it is unique?

Right now, I have this added:
```
if (Directory.Exists(tempPath))
{
   tempPath = Path.Combine(tempPath, Guid.NewGuid().ToString());
}

Though, I'm not sure if this is even the correct way to go about fixing it.
```

**Comment by justinormont** [comment]
@jwood803: I think there's still a race condition since there's a delay before the `tempPath` is used. So two processes could both check and find that their path is ok to use, then both begin using the same folder.

I think you're on the right track adding a GUID. I haven't worked thru the code, but just always adding a GUID should work.

**Comment by jwood803** [comment]
@justinormont Good point! Always adding the GUID should help solve the issue. PR will come shortly.

**Comment by mstfbl** [comment]
This issue will automatically be closed by PR #4645 .
